### PR TITLE
ci: improve Go caching

### PR DESCRIPTION
**Pull request by ammario**
... Amongst various other minor improvements.

### CODE REVIEW

File: .github/workflows/ci.yaml [hunk 1]
```diff
@@ -315,15 +309,8 @@ jobs:
...
-      - uses: actions/upload-artifact@v3
-        if: success() || failure()
-        with:
-          name: gotests-postgres.xml
-          path: ./gotests.xml
-          retention-days: 30
-
       - uses: ./.github/actions/upload-datadog
-        if: always()
+        if: success() || failure()
```
Comment (matifali) [inline .github/workflows/ci.yaml:313]: Is this to skip uploading the skipped jobs?

File: .github/actions/setup-go/action.yaml [hunk 2]
```diff
@@ -1,36 +1,62 @@
...
+        # We do our own caching for implementation clarity.
+        cache: false
+        go-version: ${{ inputs.version }}
+
+    - name: Get cache dirs
+      shell: bash
+      run: |
+        set -x
+        echo "GOMODCACHE=$(go env GOMODCACHE)" >> $GITHUB_ENV
+        echo "GOCACHE=$(go env GOCACHE)" >> $GITHUB_ENV
```
Comment (deansheather) [inline .github/actions/setup-go/action.yaml:31]: This doesn't work on Windows runners

File: .github/actions/setup-go/action.yaml [hunk 3]
```diff
@@ -1,36 +1,62 @@
...
+    # We split up GOMODCACHE from GOCACHE because the latter must be invalidated
+    # on code change, but the former can be kept.
+    - name: Cache $GOMODCACHE
+      uses: buildjet/cache@v3
+      with:
+        path: |
+          ${{ env.GOMODCACHE }}
+        key: gomodcache-${{ runner.os }}-${{ hashFiles('**/go.sum') }}-${{ github.job }}
+        restore-keys: |
+          gomodcache-${{ runner.os }}-${{ hashFiles('**/go.sum') }}-
```
Comment (deansheather) [inline .github/actions/setup-go/action.yaml:42]: GOCACHE also has the github.job one listed

File: .github/actions/setup-go/action.yaml [hunk 4]
```diff
@@ -1,36 +1,62 @@
...
+          ${{ env.GOCACHE }}
         # Job name must be included in the key for effective
         # test cache reuse.
-        key: go-${{ runner.os }}-${{ github.job }}-${{ hashFiles('**/*.go', 'go.**') }}
+        key: gocache-${{ runner.os }}-${{ github.job }}-${{ hashFiles('**/*.go', 'go.**') }}
         restore-keys: |
-          go-${{ runner.os }}-${{ github.job }}-
-          go-${{ runner.os }}-
-          go-
+          gocache-${{ runner.os }}-${{ github.job }}-
```
Comment (deansheather) [inline .github/actions/setup-go/action.yaml:54]: Hyphen at the end?

File: .github/actions/setup-go/action.yaml [hunk 5]
```diff
@@ -1,36 +1,62 @@
...
+          gocache-${{ runner.os }}-
 
     - name: Install gotestsum
-      uses: jaxxstorm/action-install-gh-release@v1.10.0
-      with:
-        repo: gotestyourself/gotestsum
-        tag: v1.9.0
+      shell: bash
+      run: go install gotest.tools/gotestsum@latest
```
Comment (deansheather) [inline .github/actions/setup-go/action.yaml:59]: Does not work on Windows AFAIK

File: .github/workflows/ci.yaml [hunk 6]
```diff
@@ -100,49 +100,45 @@ jobs:
 
       - uses: ./.github/actions/setup-go
 
-      # Check for any typos!
+      - uses: ./.github/actions/setup-node
+
+      - name: Get golangci-lint cache dir
+        run: |
+	        go install github.com/golangci/golangci-lint/cmd/golangci-lint@v1.53.2
```
Comment (deansheather) [inline .github/workflows/ci.yaml:107]: Bad indenting

File: .github/workflows/ci.yaml [hunk 7]
```diff
@@ -565,7 +547,7 @@ jobs:
 
   required:
     runs-on: ubuntu-latest
-    needs: [gen, test-go, test-go-pg, test-go-race, test-js]
+    needs: [gen, fmt, lint, test-go, test-go-pg, test-go-race, test-js]
```
Comment (deansheather) [inline .github/workflows/ci.yaml:550]: Please add these checks to the echo statements below

File: Makefile [hunk 8]
```diff
@@ -402,11 +402,17 @@ else
...
 .PHONY: lint
 
+lint/ts:
+	cd site
+	yarn && yarn lint
+.PHONY: lint/ts
+
 lint/go:
 	./scripts/check_enterprise_imports.sh
+	go install github.com/golangci/golangci-lint/cmd/golangci-lint@v1.53.2
```
Comment (deansheather) [inline Makefile:415]: Can this be skipped if golangci-lint is in PATH?

File: enterprise/cli/root_internal_test.go [hunk 9]
```diff
@@ -3,9 +3,10 @@ package cli
 import (
 	"testing"
 
-	"github.com/coder/coder/cli"
 	"github.com/stretchr/testify/require"
```
Comment (deansheather) [inline enterprise/cli/root_internal_test.go:7]: The coder/coder imports should be grouped

**Review by deansheather — COMMENTED** [review]
The Setup Go workflow needs to work on Windows

File: .github/actions/setup-go/action.yaml (hunk 2 above)
Comment (ammario) [reply .github/actions/setup-go/action.yaml:31]: Huh?

File: .github/actions/setup-go/action.yaml (hunk 4 above)
Comment (ammario) [reply .github/actions/setup-go/action.yaml:54]: That was on purpose. It's how I've seen it done on a few docs. I guess it clearly symbolizes the end of a token.

File: .github/actions/setup-go/action.yaml (hunk 5 above)
Comment (ammario) [reply .github/actions/setup-go/action.yaml:59]: why?

File: Makefile (hunk 8 above)
Comment (ammario) [reply Makefile:415]: Why do it though? It runs almost instantly if already installed.

File: .github/actions/setup-go/action.yaml (hunk 3 above)
Comment (ammario) [reply .github/actions/setup-go/action.yaml:42]: I added a comment that may help this make more sense

**Comment by ammario** [comment]
@deansheather the windows job seems to run fine?

**Comment by ammario** [comment]
Force merging as folks are struggling with CI today.

**Pull request merged**